### ISSUE TITLE
Issue with the multiple use of the index (with STAR command)

### ISSUE THREAD

**Issue by cfpcxs** [nextflow-io/hack17-varcall#1]
Hello,

I'm new in Nextflow use and I try to make a pipeline for RNA_Seq analysis in Nextflow.
It looks like it works for different sample of fastq files until the Alignment steps.
From the Alignment steps it doesn't work anymore for different sample: it's just run the script for one sample and finish.
I use STAR for building the index and for the alignment step, in a pretty similar way than you in your script: https://github.com/nextflow-io/hack17-varcall/blob/master/solutions/fat_floor.adoc .
When I need to build the index (with STAR), I put the directory of the index in a channel. In the next process (Alignment with STAR), I take back the directory of the index from this channel and use it in the STAR command. 
When I have different fastq files like, for example: "12CP.fastq", "6HP.fastq", "8HPT.fastq" of the Potatoes species, all process before Alignment run with all the samples and produce all the files. But from the BuildIndex and Alignment, it's running just for one sample and finish the pipeline without returning an error, like if all it's ok. Maybe it's because we can't use the same directory index for multiple STAR Alignment command ?
So I would know if your Nextflow script run for different sample? Or do you have the same problem?

Thank you for your attention
Best regards

Carole

**Comment by pditommaso** [comment]
I think this is happening because you may have declared the build index process input as a channel. Since you want to build index just once you need to specify that input just as file reference. 

Look for example how is declare [genome_file](https://github.com/CRG-CNAG/CalliNGS-NF/blob/master/main.nf#L120) in this process that's doing exactly what you are looking for.

**Comment by cfpcxs** [comment]
Hello,

Thank you for your reply!
I think I made what you are saying, I have declared the build index input as a channel. Here is the part of my script that works for only one sample. Do you think I did something wrong? 

Thank you for your attention

```nextflow
process buildIndex {
    executor "lsf"
    module "STAR/2.5.3a"
    publishDir params.outdir, mode: 'copy'

    input:
    file(genome) from genome_file 
    file(annot) from annotation_file

    output:
    file(indexDir) into genome_indexD 
    file(annot) into annotation_file2

"""
mkdir indexDir
STAR --runMode genomeGenerate --genomeDir indexDir --genomeFastaFiles ${genome} --sjdbGTFfile ${annot} --sjdbGTFtagExonParentTranscript Parent 
"""
 } 



/* Alignment with Star: Single Pass Mode: input are genome, genome index and fastqtrimmed data, output are BAM file*/

process alignment {
    executor "lsf"
    module "STAR/2.5.3a"
    module "picard-tools"
    publishDir params.outdir, mode: 'copy'

    input:
    file(index) from genome_indexD
    set pair_id, file(treads) from TrimmomaticPaired3 

    output:
    set pair_id, file("BAMD.*") into bam 
    
"""
mkdir BAMD.${pair_id}
STAR --genomeDir ${index} --outFileNamePrefix BAMD.${pair_id}/${pair_id}_ --readFilesCommand zcat --readFilesIn ${pair_id}.trimmomatic_1P.fq.gz ${pair_id}.trimmomatic_2P.fq.gz --chimSegmentMin 15 --chimJunctionOverhangMin 15 --alignMatesGapMax 20000 --alignIntronMax 20000 --outSAMtype BAM Unsorted --outSAMprimaryFlag AllBestScore --outSAMstrandField intronMotif --outFilterIntronMotifs RemoveNoncanonicalUnannotated --quantMode GeneCounts --outQSconversionAdd -31
"""
}
```

**Comment by pditommaso** [comment]
> I have declared the build index input as a channel

No, it shouldn't be declared like a channel. It should be declared something like:

```
genome_file = file(params.genome)
```

**Comment by Juke34** [comment]
doing `file(index) from genome_indexD.collect()` should work
The .collect() operator should mean that the channel isn't consumed and is used for every task of the process.
